Sources/NIOPosix/Bootstrap.swift

```diff
@@ -3169,6 +3169,324 @@ public final class NIOPipeBootstrap {
+    ///         when the `Channel` becomes inactive. You _must not_ do any further operations `inputDescriptor` or
+    ///         `outputDescriptor`, including `close`.
+    ///         If this method throws, you still own the file descriptors and are responsible for
+    ///         closing them.
+    ///
+    /// - Parameters:
+    ///   - fileDescriptor: The _Unix file descriptor_ for the input & output.
+    ///   - channelInitializer: A closure to initialize the channel which must return the handler that is used for negotiating
+    ///   the protocol.
+    /// - Returns: The protocol negotiation result.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func withInputOutputDescriptor<Handler: NIOProtocolNegotiationHandler>(
+        _ fileDescriptor: CInt,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Handler>
+    ) async throws -> Handler.NegotiationResult {
+        let inputFD = fileDescriptor
+        let outputFD = try! Posix.dup(descriptor: fileDescriptor)
+
+        do {
+            return try await self.withPipes(
+                inputDescriptor: inputFD,
+                outputDescriptor: outputFD,
+                channelInitializer: channelInitializer
+            )
+        } catch {
+            try! Posix.close(descriptor: outputFD)
+            throw error
+        }
+    }
+    
+    /// Create the `PipeChannel` with the provided input and output file descriptors.
+    ///
+    /// The input and output file descriptors must be distinct. If you have a single file descriptor, consider using
+    /// `ClientBootstrap.withConnectedSocket(descriptor:)` if it's a socket or
+    /// `NIOPipeBootstrap.withInputOutputDescriptor` if it is not a socket.
+    ///
+    /// - Note: If this method returns normally, SwiftNIO will close `inputDescriptor` and `outputDescriptor`
+    ///         when the `Channel` becomes inactive. You _must not_ do any further operations `inputDescriptor` or
+    ///         `outputDescriptor`, including `close`.
+    ///         If this method throws, you still own the file descriptors and are responsible for
+    ///         closing them.
+    ///
+    /// - Parameters:
+    ///   - inputDescriptor: The _Unix file descriptor_ for the input (ie. the read side).
+    ///   - outputDescriptor: The _Unix file descriptor_ for the output (ie. the write side).
+    ///   - channelInitializer: A closure to initialize the channel which must return the handler that is used for negotiating
+    ///   the protocol.
+    /// - Returns: The protocol negotiation result.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    @_spi(AsyncChannel)
+    public func withPipes<Handler: NIOProtocolNegotiationHandler>(
+        inputDescriptor: CInt,
+        outputDescriptor: CInt,
+        channelInitializer: @escaping @Sendable (Channel) -> EventLoopFuture<Handler>
+    ) async throws -> Handler.NegotiationResult {
```

Comment:
Can the body of this be written in terms of the variant that returns `Output`? There's a lot of duplication.